### PR TITLE
Add response header for tracking.

### PR DESCRIPTION
**Pull request by T0mmyliu**

### CODE REVIEW

File: tests/test_tracking.py [hunk 1]
```diff
@@ -1,17 +1,16 @@
 # -*- coding: utf-8 -*-
 
-from __future__ import absolute_import
```
Comment (wooparadog) [inline tests/test_tracking.py]: Why is this deleted?

File: tests/test_tracking.py [hunk 2]
```diff
@@ -28,7 +27,10 @@
 from thriftpy.transport import TServerSocket, TBufferedTransportFactory, \
     TTransportException, TSocket
 from thriftpy.protocol import TBinaryProtocolFactory
-
+from compatible.version_2.tracking \
```
Comment (wooparadog) [inline tests/test_tracking.py]: better style can be

```python
from a.b.c import (
    a,
    b,
    c,
)
```

File: tests/test_tracking.py [hunk 3]
```diff
@@ -28,7 +27,10 @@
...
+    import TTrackedProcessor as old_TTrackedProcessor, \
```
Comment (wooparadog) [inline tests/test_tracking.py]: Do not use `old_` as prefix.  `import TTrackedProcessor as TTrackedProcessorV2` is a lot better.

File: tests/test_tracking.py [hunk 4]
```diff
@@ -45,18 +47,29 @@ def _get_port():
...
 
 class SampleTracker(TrackerBase):
     def record(self, header, exception):
         db = dbm.open(db_file, 'w')
+        key = "%s:%s" % (header.request_id, header.seq)
+        db[key.encode("ascii")] = pickle.dumps(header.__dict__)
+        db.close()
+
 
+class OldTracker(old_TrackerBase):
```
Comment (wooparadog) [inline tests/test_tracking.py]: Different trackers of different versions should also be imported from the module of that version.

File: tests/test_tracking.py [hunk 5]
```diff
@@ -265,7 +299,7 @@ def test_tracker_chain(server, server1, server2, dbm_db, tracker_ctx):
     with client() as c:
         with SampleTracker.add_meta(**test_meta):
             c.remove("jane")
-            c.hello("yes")
+        c.hello("yes")
```
Comment (wooparadog) [inline tests/test_tracking.py]: why is this?

File: tests/test_tracking.py [hunk 6]
```diff
@@ -279,9 +313,7 @@ def test_tracker_chain(server, server1, server2, dbm_db, tracker_ctx):
     assert len(set([i["request_id"] for i in headers])) == 2
 
     seqs = [i["seq"] for i in headers]
-    metas = [i["meta"] for i in headers]
     assert seqs == ['1', '1.1', '1.1.1', '1.1.2', '2']
-    assert metas == [test_meta] * 5
```
Comment (wooparadog) [inline tests/test_tracking.py]: Why removing asserts here?

File: tests/test_tracking.py [hunk 7]
```diff
@@ -372,3 +388,136 @@ def test_counter(server, dbm_db, tracker_ctx):
...
+
+'''
+The following 7 test cases tests the backward compatibility,
+there are three kinds of client/server : not_tracked
+old_tracked(only support request header not support response header),
+new_tracked(support request and response header)
+'''
+
+
+def test_not_tracked_client_new_tracked_server(server):
```
Comment (wooparadog) [inline tests/test_tracking.py]: naming could be better:

`test_not_tracked_client_new_tracked_server` => `test_native_client_call_v2_tracking_server`

File: thriftpy/contrib/tracking/__init__.py [hunk 8]
```diff
@@ -94,13 +99,21 @@ def _send(self, _api, **kwargs):
         self.send_start = int(time.time() * 1000)
         super(TTrackedClient, self)._send(_api, **kwargs)
 
+    def _recv(self, _api):
+        if self._upgraded \
+                and self.check_support(TrackerVersion.support_response_header):
+            self.response_header = track_thrift.ResponseHeader()
```
Comment (wooparadog) [inline thriftpy/contrib/tracking/__init__.py]: There's no need to bind `response_header` to `self`

File: thriftpy/contrib/tracking/__init__.py [hunk 9]
```diff
@@ -57,6 +58,8 @@ def __init__(self, tracker_handler, *args, **kwargs):
 
         self.tracker = tracker_handler
         self._upgraded = False
+        self.client_version = TrackerVersion.support_response_header
```
Comment (wooparadog) [inline thriftpy/contrib/tracking/__init__.py]: This could be constant.

File: thriftpy/contrib/tracking/__init__.py [hunk 10]
```diff
@@ -123,13 +136,20 @@ def _req(self, _api, *args, **kwargs):
...
+
 
 class TTrackedProcessor(TProcessor):
     def __init__(self, tracker_handler, *args, **kwargs):
         super(TTrackedProcessor, self).__init__(*args, **kwargs)
-
         self.tracker = tracker_handler
         self._upgraded = False
+        self.add_response_header = False
+        self.server_version = TrackerVersion.support_response_header
```
Comment (wooparadog) [inline thriftpy/contrib/tracking/__init__.py]: This can also be constant..

File: thriftpy/contrib/tracking/__init__.py [hunk 11]
```diff
@@ -195,7 +219,20 @@ def _do_process(self, iprot, oprot, api, seqid, result, call):
             self.handle_exception(e, result)
 
         if not result.oneway:
+            if self._upgraded and self.add_response_header:
+                response_header = track_thrift.ResponseHeader()
+                self.tracker.gen_response_header(response_header)
+                response_header.write(oprot)
+
+            if self.check_support(TrackerVersion.support_response_header):
```
Comment (wooparadog) [inline thriftpy/contrib/tracking/__init__.py]: there is no need to check whether we should send response header in every rpc.

File: thriftpy/contrib/tracking/__init__.py [hunk 12]
```diff
@@ -195,7 +219,20 @@ def _do_process(self, iprot, oprot, api, seqid, result, call):
...
+                self.add_response_header = True
             self.send_result(oprot, api, result, seqid)
 
+    def check_support(self, version):
```
Comment (wooparadog) [inline thriftpy/contrib/tracking/__init__.py]: Identical method could be extracted to one function.

File: thriftpy/contrib/tracking/tracker.py [hunk 13]
```diff
@@ -10,14 +10,23 @@
...
 class TrackerBase(object):
     def __init__(self, client=None, server=None):
         self.client = client
         self.server = server
+        self.version = TrackerVersion.support_response_header
 
     def handle(self, header):
         ctx.header = header
-        ctx.counter = 0
+        ctx.counter = TrackerVersion.default
```
Comment (wooparadog) [inline thriftpy/contrib/tracking/tracker.py]: The `0` here doesn't mean `TrackerVersion.default`.

Comment (wooparadog) [inline thriftpy/contrib/tracking/tracker.py]: The `0` here doesn't mean `TrackerVersion.default`.

Comment (wooparadog) [inline thriftpy/contrib/tracking/tracker.py]: The `0` here doesn't mean `TrackerVersion.default`.

File: thriftpy/contrib/tracking/__init__.py (hunk 10 above)
Comment (T0mmyliu) [inline thriftpy/contrib/tracking/__init__.py]: self.add_response_header = False
self.server_version = TrackerVersion.support_response_header

Maybe... the second should be constant
The first should not.

File: thriftpy/contrib/tracking/tracker.py (hunk 13 above)
Comment (T0mmyliu) [inline thriftpy/contrib/tracking/tracker.py]: stupid error...
